### PR TITLE
[HLSL] Parameter modifier parsing and AST

### LINKED ISSUE #59849: Parameter InputModifier like in/out/inout
Support InputModifier like in/out/inout
https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-function-parameters

### PR DESCRIPTION
**Pull request by llvm-beanz**
This change implements parsing for HLSL's parameter modifier keywords `in`, `out` and `inout`. Because HLSL doesn't support references or pointers, these keywords are used to allow parameters to be passed in and out of functions.

This change only implements the parsing and AST support. In the HLSL ASTs we represent `out` and `inout` parameters as references, and we implement the semantics of by-value passing during IR generation.

In HLSL parameters marked `out` and `inout` are ambiguous in function declarations, and `in`, `out` and `inout` may be ambiguous at call sites.

This means a function may be defined as `fn(in T)` and `fn(inout T)` or `fn(out T)`, but not `fn(inout T)` and `fn(out T)`. If a funciton `fn` is declared with `in` and `inout` or `out` arguments, the call will be ambiguous the same as a C++ call would be ambiguous given declarations `fn(T)` and `fn(T&)`.

Fixes #59849

### CODE REVIEW

**Review by python3kgae — APPROVED** [review]
Maybe a verify test for use rvalue for inout parameter.

File: clang/lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -5559,7 +5562,6 @@ bool Parser::isTypeSpecifierQualifier() {
   case tok::kw___read_write:
   case tok::kw___write_only:
   case tok::kw___funcref:
-  case tok::kw_groupshared:
     return true;
```
Comment (bogner) [inline clang/lib/Parse/ParseDecl.cpp:5565]: This is technically an unrelated bugfix. Not sure it's worth committing it separately, but maybe we should add a test that these keywords are not accepted in C++ mode similar to how we test some of the opencl keywords (see `clang/test/SemaOpenCLCXX/private-access-specifier.cpp`)

File: clang/test/SemaHLSL/parameter_modifiers.hlsl [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+// Alternating `inout` and `in out` spellings between declaration and
+// definitions is fine since they have the same semantic meaning.
+void fn7(inout float f);
+void fn7(in out float f) {}
+
+void fn8(in out float f);
+void fn8(inout float f) {}
+
+// These two declare two different functions (although calling them will be
+// ambiguous). This is equivalent to declaring a functiion that takes a
```
Comment (bogner) [inline clang/test/SemaHLSL/parameter_modifiers.hlsl:53]: Typo. s/functiion/function/

**Review by bogner — APPROVED** [review]
